### PR TITLE
[Azure Search] Add support for Complex Types to the .NET SDK

### PR DESCRIPTION
**Pull request by bitsforthought**
These changes add support for Complex Types to the SDK. Scenarios covered include:
- Defining complex fields procedurally with the `Field` class
- Defining complex fields and non-string collections declaratively with `FieldBuilder`
- Indexing documents containing complex fields using statically-typed model classes or the `Document` class (property bag).
- Retrieving documents containing complex fields using statically-typed model classes or the `Document` class via the `Get`, `Search`, and `Suggest` APIs.

FYI @Yahnoosh @ishansrivastava90 @updixit @arv100kri

### CODE REVIEW

**Review by dsgouda — COMMENTED** [review]
LGTM, @brjohnstmsft are you planning to bump Nuget  versions in a separate PR?

**Comment by brjohnstmsft** [comment]
@dsgouda We bumped the version several months ago at the beginning of this development arc. The version of the SDK with all these changes hasn’t shipped yet.

**Comment by dsgouda** [comment]
> 
> 
> @dsgouda We bumped the version several months ago at the beginning of this development arc. The version of the SDK with all these changes hasn’t shipped yet.

Thanks for confirming, will approve once this is ready to merge

**Comment by brjohnstmsft** [comment]
Build failures unrelated to my changes.

**Comment by dsgouda** [comment]
> 
> 
> Build failures unrelated to my changes.

Taking a look

**Comment by dsgouda** [comment]
@brjohnstmsft Please update the search-preview branch with latest changes from master, this should fix the build failures.

**Comment by brjohnstmsft** [comment]
@dsgouda I can try that, but if that fixes it, then it must be an intermittent problem since we were able to successfully merge from master a few weeks ago.

**Comment by dsgouda** [comment]
> 
> 
> @dsgouda I can try that, but if that fixes it, then it must be an intermittent problem since we were able to successfully merge from master a few weeks ago.

The last time search-preview branch was updated the tests must have been broken and we had disabled them in our Jenkins CI. Now that they are fixed, we have reenabled them which is why the branch must now be updated.

**Comment by brjohnstmsft** [comment]
@dsgouda Got it, thanks. I'll merge from master to resolve this.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Indexes/Models/Field.Customization.cs
Comment (shmed) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Indexes/Models/Field.Customization.cs]: very small nit (and subjective too): The name of this function doesnt seem to accurately explain what it does. It doesn't really "throw if argument out of range", it "throw's argument out of range IF parameter is true". The description of the function explains what it does, but by looking at its usage here I spent some time wondering how it was able to know if something was out of range.

**Review by bitsforthought — COMMENTED** [review]
Update in a subsequent PR

**Review by shmed — APPROVED** [review]

**Review by natinimni — APPROVED** [review]
LGTM

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/FieldBuilderTests.cs
Comment (brjohnstmsft) [reply src/SDKs/Search/DataPlane/Search.Tests/Tests/FieldBuilderTests.cs]: Decided not to do this in case it would interfere with some scenario we can't foresee.

**Pull request merged**
